test/albumController.test.js

```diff
@@ -0,0 +1,105 @@
+const { app, connectDB } = require('../src/expressServer'); // Import your Express app here
+const { StatusCodes } = require('http-status-codes');
+const request = require('supertest');
+const { MongoMemoryServer } = require('mongodb-memory-server');
+const Album = require('../src/models/Album');
+
+const {
+  getAllAlbums,
+  getSingleAlbum,
+  getAlbumWithAllUsersWhoPurchasedIt,
+  getFilteredAlbums,
+} = require('../src/controllers/albumController');
+
+let server;
+let mongooseConnection;
+let mongodb;
+
+beforeAll(async () => {
+  mongodb = await MongoMemoryServer.create();
+  const url = mongodb.getUri();
+  process.env.MONGO_URL = url;
+  mongooseConnection = await connectDB(url);
+  server = await app.listen(8001);
+});
+
+afterAll(async () => {
+  await server.close();
+  await mongooseConnection.disconnect();
+  await mongodb.stop();
+});
+
+describe('AlbumController API Tests', () => {
+  const mockAlbumData = {
+    albumName: 'Test Album',
+    artistName: 'Test Artist',
+    price: 9.99,
+    spotifyUrl: 'https://api.spotify.com/v1/albums/',
+  };
+
+  beforeEach(async () => {
+    await Album.deleteMany({});
+    await Album.create(mockAlbumData);
+  });
+
+  it('should test the getAllAlbums endpoint - Success Case', async () => {
+    const response = await request(app).get('/api/v1/albums');
+    expect(response.status).toBe(StatusCodes.OK);
+    expect(response.body).toHaveProperty('albums');
+    expect(response.body.albums).toHaveLength(1);
+  });
+
+  it('should test the getAllAlbums endpoint - Error Case (Unauthorized)', async () => {
+    const response = await request(app).get('/api/v1/albums');
+    expect(response.status).toBe(StatusCodes.UNAUTHORIZED);
+  });
```

Comment:
I have tried to do as you explained , I have changed it for varyfing if the endpoints returns  an empty list, could you please have a look? commit 23b4f0ecdf3ecb19bdc226ec04792096dfec04e0

Reply:
The `beforeEach` step creates an album though right; so we need to delete that album in the test:

You have:

```
  it('should return an empty list if there are no albums in the database', async () => {
    const response = await request(app).get('/api/v1/albums');
    expect(response.status).toBe(StatusCodes.UNAUTHORIZED);
    expect(response.body).toEqual([]);
  });
```

But (1) this won't be an "Unauthorized" status code, it'll be a 200 because we don't check for logged in users on this endpoint. And (2) we need to first delete the albums before we can do the test so:

```
  it('should return an empty list if there are no albums in the database', async () => {
    await Album.deleteMany({});
    const response = await request(app).get('/api/v1/albums');
    expect(response.status).toBe(StatusCodes.OK);
    expect(response.body).toEqual([]);
  });
```

Reply:
Thank you , ok I understood, commit 6ebe135cf632e360db1ddde5a71323b08b647acf